openedx/core/djangoapps/user_authn/views/register.py

```diff
@@ -568,6 +569,13 @@ def post(self, request):
 
         data = request.POST.copy()
         self._handle_terms_of_service(data)
+        try:
```

Comment:
I'm a little concerned about how we maintain the contract here with `form_data`, since it seems like the LMS is free to remove a field at any time without the filter needing to be updated, which would potentially cause filters to break downstream. I'm not sure how to address that given the optional nature of a lot of form data. I'm not going to block this PR on it, since I think it's more important to get this out for people to try, and it's better than what exists now. But I am curious what your thoughts on it are these days.

Reply:
Formally, we haven't had any discussion about this specific case. But what I think is, why don't we take a _screenshot_ of the form_data contents as they are right now? And use that as v1. Then, if the form's contents change, we update the filter.

What do you think? @ormsbee @felipemontoya

Reply:
In this scenario, i.e. if you bump up the version of the filter from v1 to v2, how do those two interoperate? Do they interoperate?

What I mean is, if you bump up a filter to v2, will support be immediately dropped for v1? If not, then how do you work in a mixed environment of some v1 filters and some v2 filters.

Reply:
Following the [Open edX filters naming and versioning](https://github.com/eduNEXT/openedx-filters/tree/main/docs/decisions) ADR, they must interoperate:
- Open edX core developers wanting to deprecate and eventually remove filters.

This is what I imagine it'd be:
```
# data.keys -> ["username", "email", "name", "password", "mailing_address", ""]
data = PreRegisterFilter.run(form_data=data)  # Filter expects data.keys -> ["username", "email", "name", "password"]
data = PreRegisterFilterV2.run(form_data=data) # Filter expects data.keys -> ["username", "password"]
data = PreRegisterFilterV3.run(form_data=data) # Filter expects data.keys -> ["username"]
```
Filters, by definition, must return what they receive (the whole data.keys), but they'll be able to use just the subset that the filter defines. I imagine this evolution happening because we decide that the filter can't modify some fields :thinking:

This is just me throwing some ideas. I'm thrilled to hear what you think

Reply:
@mariajgrimaldi:

> Formally, we haven't had any discussion about this specific case. But what I think is, why don't we take a screenshot of the form_data contents as they are right now? And use that as v1. Then, if the form's contents change, we update the filter.

So we make the current fields a part of the filter's contract, and if edx-platform changes the field names for some reason, we do the translation at the filter level so that any pipeline steps that have been written are unaffected? That sounds good to me. The only question I'd have then is how we detect a break in the contract from edx-platform testing–i.e. if I change a field name right now, will it automatically get caught by a filter-related test breaking in CI?

Reply:
I agree with the sentiment that we should not block the PR on the issue of maintaining the contract that `form_data` provides. I'm actually going to propose the counter argument. Should we at all try lock in the keys in form_data dict and maintain them? I think not.

There is always going to be some degree of flexibility due to the optional fields which will make this very difficult if at all possible. From my perspective, the Filter at the student registration is the key piece of extension that allows any sort of registration flow to be handled. A developer could now remove or dramatically alter the form at the registration page and send any sort of info, which the filter will receive and "normalize" into something that the core platform can use to create the final user.

E.g:
- a custom MFE for registration gathers names in spanish customs (2 names, two last names), national id, and phone (I've seen this kind of requirement many times).
- the filter processes this dict and creates a completely new `form_data` that is consistent with the openedx user models and views. Fullname, meta fields for ID and all that. Creates a password since it was not included and emails it to the user (horrible experience, but I have also have had this requested from us).

Filters, specially one like the filter before a student registers, are going to have some risk of breaking since we are passing in memory objects to them. The same would be true for a filter that receives a user object or enrollment object. Eventually it will have to deal with certain methods of the object changing. I think we have to embrace that and learn to write filters that are delicate in the way the handle the live data that is passed to them.

Reply:
I think I'm on the side of _not enforcing the `form_data` contract_, it's up to filter developers to make sure their filters return broken data to the platform (it's similar to Django plugin apps - we have more freedom, but a plugin can easily break the LMS if hooked wrong).

Reply:
@ormsbee has any of this conversation swayed you to not fixing in place the content of form_data?
This is probably the last outstanding request to change the current form of this PR. If the issue is not blocking for you, would you agree that we go ahead with this?

Reply:
@felipemontoya: Yeah, it's not a blocker.